### PR TITLE
fix(runtime): preserve launch policy and failed attempts during continuation

### PR DESCRIPTION
**Pull request by aviggiano**
A native continuation can fail before the provider runs because it loses the governance authority admitted at launch. A stopped workflow reset can also overwrite a failed executor's mutable attempt detail before that terminal occurrence has been recorded; the runner then finishes while status and stats refuse the missing history.

Restore only the authenticated governance path from the original protected execution snapshot when a continuation will actually launch. Preserve the original provider policy and native runner environment. Active attachment remains a no-op. Ordinary no-reset continuation retains native tolerance for missing or malformed link journals. A current run that claims a seal but has lost it refuses before reset or provider execution. Unsealed legacy workflows and authenticated original sealed plans from before governance was introduced retain their existing path; partial policy claims refuse.

Before an actual reset of a stopped workflow, preserve recordable failed local executor occurrences using the existing selected-agent authority, immutable digest, reconciliation and ledger semantic checks. A stopped reset of a current sealed run requires valid linked/control/task evidence, including its link journal, before candidate selection; unavailable authority refuses before reset even if no executed failure would qualify. Validate every candidate and recheck runner/control evidence before the durable append and scheduler mutation. Repeated checkpoint/reset attempts retain the same immutable rows. Successful-result admission, failures before agent selection, active forced resets and ordinary no-reset continuation keep their existing behavior. This does not perform full synchronization, finalize artifacts, or rewrite old history.

The external generated-run integrations use the pinned real runner and deterministic local provider executable: failure without prior observation, supported reset, successful replacement, real verifier completion, then repeated built CLI status/stats. Both ordinary and refreshed reset pass without a manual governance-path input. Each retains failed event 24 byte-for-byte, appends successful event 48 once despite reused physical attempt 1, and gives four clean CLI observations reporting two executed attempts and one retry. Committed regressions cover ordinary and refreshed continuation, malformed authority, interrupted reset, incompatible existing rows and no-op/legacy boundaries.

Validation: Node 24 CLI/dependency build and runtime test compilation; 17 focused combined runtime regressions pass; strict changed-code lint, formatting and diff checks pass. Independent source review covers both fixes and their callback ordering. The original main fixture reproduces the missing policy; a separately labeled original-policy compensation isolates the lost failure history. All integration providers are deterministic local executables.

Native continuation retains its existing execution model. Policy authentication detects missing or changed original authority before launch; it does not introduce whole-snapshot descriptor confinement for native execution. Deliberate same-owner replacement of a read-only generation after authentication remains outside this repair; normal publication and refresh retain the original generation.

Prevents #1099 for recordable failed executor occurrences at supported stopped-run resets. Already unrecorded superseded history remains refused; this change does not invent its missing authority.

<h3>Greptile Summary</h3>

This PR restores sealed data-governance policy for native continuations and checkpoints failed local executor attempts before stopped-run resets overwrite mutable attempt details.

- Defers continuation-policy preparation until after active-attachment detection.
- Authenticates governance policy against the original control generation.
- Preserves recordable failed attempts before retry or reset commands.
- Adds legacy, tampering, interruption, reconciliation, and idempotency regressions.

One execution-boundary gap remains: governance authentication produces a pathname rather than an open-file or snapshot capability, leaving a replacement window before the resumed child reads the policy.

<h3>Confidence Score: 3/5</h3>

The PR should not merge until authenticated governance remains bound to the exact policy bytes through the resumed child’s open.

The failed-attempt checkpoint is carefully validated, but the continuation-policy repair has a pathname replacement window between successful authentication and provider execution, so launch governance can still be substituted.

**Files Needing Attention:** packages/runtime/src/workflow-integrity.ts, packages/runtime/src/start-run.ts

<details open><summary><h3>Security Review</h3></summary>

The governance policy is authenticated before launch, but the authenticated object is reduced to a pathname. Because the resumed child opens that path later without seal verification or descriptor pinning, same-owner replacement of the snapshot path can substitute policy after authentication.
</details>

<details><summary><h3>Important Files Changed</h3></summary>

| Filename | Overview |
|----------|----------|
| packages/runtime/src/smithers.ts | Defers continuation preparation until after active attachment and invokes failure preservation before stopped resets. |
| packages/runtime/src/start-run.ts | Wires authenticated governance restoration and the stopped-reset attempt checkpoint into continuation submission. |
| packages/runtime/src/workflow-integrity.ts | Authenticates sealed governance content but returns an unpinned pathname that is reopened after authentication. |
| packages/runtime/src/workflow-sync.ts | Adds validated, idempotent preservation of failed local executor occurrences before reset. |
| packages/runtime/test/runtime.test.ts | Adds broad regression coverage for governance restoration, legacy behavior, reset preservation, interruption, and malformed evidence. |
| packages/runtime/test/fixtures/pre-governance-run-plan.ts | Adds a canonical pre-governance run-plan fixture for legacy continuation coverage. |

</details>

<details><summary><h3>Sequence Diagram</h3></summary>

```mermaid
sequenceDiagram
    participant Caller as Resume caller
    participant Runtime as Runtime lifecycle
    participant Seal as Sealed snapshot
    participant Ledger as Attempt ledger
    participant Smithers as Smithers child

    Caller->>Runtime: resume / retry / reset
    Runtime->>Smithers: inspect current run
    alt active attachment
        Runtime-->>Caller: already running
    else continuation launches
        Runtime->>Seal: verify seal, plan, and governance digest
        Seal-->>Runtime: governance pathname
        opt stopped reset
            Runtime->>Smithers: read events and attempt authority
            Runtime->>Ledger: validate and append failed attempts
            Runtime->>Smithers: timetravel/reset
        end
        Runtime->>Smithers: launch with governance pathname
        Smithers->>Seal: open policy by pathname
    end
```
</details>

<a href="https://app.greptile.com/ide/claude-code?prompt=Greploop%20monad-developers%2Fultrafuzz%20PR%20%231117%3A%20work%20through%20Greptile's%20open%20review%20comments%2C%20then%20keep%20reviewing%20and%20fixing%20until%20it%20comes%20back%20clean%20at%205%2F5%20with%20zero%20unresolved%20comments.%0AStart%20by%20reading%20the%20comments%20off%20the%20PR%20itself.%20On%20GitHub%2C%20use%20paginated%20%60gh%20api%20graphql%60%20to%20query%20%60pullRequest.reviewThreads%60%20with%20each%20thread's%20%60isResolved%60%20value%20and%20inline%20%60comments%60.%20%60gh%20pr%20view%20--comments%60%20only%20includes%20conversation%20comments%2C%20so%20do%20not%20use%20it%20as%20the%20findings%20source.%20They%20are%20not%20listed%20in%20this%20prompt%20on%20purpose%3A%20the%20PR%20is%20current%2C%20a%20pasted%20copy%20would%20not%20be.%20Skip%20anything%20already%20resolved%2C%20and%20if%20you%20judge%20a%20comment%20wrong%2C%20say%20so%20rather%20than%20changing%20code%20to%20satisfy%20it.%0A%0APrefer%20the%20Greptile%20CLI%2C%20which%20reviews%20the%20working%20tree%20with%20no%20push%20and%20no%20CI%20run.%20Fall%20back%20to%20PUSH%20LOOP%20only%20where%20a%20step%20below%20says%20to.%0A1.%20Run%20%60command%20-v%20greptile%60.%20Missing%3A%20go%20to%20PUSH%20LOOP%2C%20and%20pitch%20the%20CLI%20at%20the%20end.%0A2.%20Run%20%60greptile%20whoami%60.%20It%20exits%200%20even%20when%20signed%20out%2C%20so%20read%20the%20output%3A%20if%20it%20says%20%22Not%20signed%20in%22%2C%20ask%20the%20user%20to%20run%20%60greptile%20login%60%20%28it%20opens%20a%20browser%2C%20do%20not%20run%20it%20yourself%29%20and%20go%20to%20PUSH%20LOOP%20with%20no%20pitch.%0A3.%20Look%20for%20%60greploop%2FSKILL.md%60%20under%20%60.agents%2Fskills%2F%60%20or%20%60.claude%2Fskills%2F%60%2C%20both%20in%20the%20repo%20root%20and%20in%20the%20user's%20home%20directory.%20Found%3A%20follow%20that%20skill%2C%20never%20pushing%20between%20rounds%2C%20and%20you%20are%20done.%0A4.%20Missing%3A%20run%20%60greptile%20skills%20list%60.%20If%20it%20fails%20as%20an%20unknown%20command%2C%20this%20CLI%20predates%20the%20skill%2C%20so%20tell%20the%20user%20to%20run%20%60greptile%20update%60%20and%20go%20to%20PUSH%20LOOP%20with%20no%20pitch.%20Otherwise%20ask%20the%20user%20before%20running%20%60greptile%20skills%20install%20greploop%60%2C%20then%20follow%20the%20skill%20it%20writes.%0A%0APUSH%20LOOP%20%28only%20when%20a%20step%20above%20sends%20you%20here%29%3A%0A%0AGREPLOOP%20ENABLED%3A%20After%20fixing%20all%20issues%2C%20commit%20and%20push%20your%20changes.%20Greptile%20will%20automatically%20re-review%20the%20PR%20via%20webhook.%20Wait%20for%20the%20review%20to%20complete%2C%20then%20read%20the%20new%20review%20comments%20and%20confidence%20score.%20If%20the%20confidence%20score%20is%20below%205%2F5%20or%20there%20are%20unresolved%20comments%2C%20fix%20those%20issues%20and%20push%20again.%20Repeat%20until%20the%20confidence%20score%20is%205%2F5%20with%20zero%20unresolved%20comments.%20Maximum%205%20iterations.%20Do%20NOT%20open%20a%20new%20PR%20%E2%80%94%20keep%20pushing%20to%20the%20same%20branch.%0A%0AIF%20YOU%20ENDED%20UP%20ON%20PUSH%20LOOP%20BECAUSE%20THE%20CLI%20WAS%20MISSING%3A%20once%20the%20loop%20is%20done%2C%20print%20the%20block%20below%20and%20nothing%20else%20about%20the%20CLI.%20Reproduce%20it%20character%20for%20character%20inside%20a%20fenced%20code%20block%2C%20so%20the%20borders%20line%20up.%20Do%20not%20reword%20it%2C%20translate%20it%2C%20summarize%20it%2C%20add%20to%20it%2C%20or%20drop%20the%20box.%20Print%20it%20once%2C%20and%20do%20not%20bring%20the%20CLI%20up%20again%20later%20in%20the%20session.%0A%0A%2B----------------------------------------------------------------------------%2B%0A%7C%20%20You%20can%20run%20greploops%20faster%20locally%20with%20our%20CLI.%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%7C%0A%7C%20%20Install%20it%20at%20https%3A%2F%2Fwww.greptile.com%2Fcli%2C%20or%20I%20can%20install%20it%20for%20you.%20%20%7C%0A%2B----------------------------------------------------------------------------%2B%0A%0AIf%20they%20take%20you%20up%20on%20it%2C%20install%20with%20%60npm%20install%20-g%20greptile%60%20%28or%20%60brew%20install%20greptileai%2Ftap%2Fgreptile%60%29%2C%20then%20%60greptile%20skills%20install%20greploop%60.%20Leave%20%60greptile%20login%60%20to%20them%2C%20it%20opens%20a%20browser.&repo=monad-developers%2Fultrafuzz&pr=1117&platform=github">Fix all with Greploop</a> <a href="https://app.greptile.com/ide/claude-code?prompt=%23%23%23%20Issue%201%0Apackages%2Fruntime%2Fsrc%2Fworkflow-integrity.ts%3A490%0A**Governance%20Path%20Is%20Unpinned**%0A%0AThe%20continuation%20authenticates%20the%20governance%20file%20but%20returns%20only%20its%20pathname.%20Smithers%20opens%20that%20path%20later%2C%20after%20the%20authentication%20checks%20have%20ended.%20If%20a%20process%20can%20modify%20the%20run-owned%20snapshot%20tree%2C%20it%20can%20replace%20the%20file%20between%20authentication%20and%20use%2C%20causing%20the%20continuation%20to%20launch%20with%20policy%20bytes%20that%20do%20not%20match%20the%20sealed%20launch%20decision.%20Keep%20the%20authenticated%20file%20or%20snapshot%20directory%20pinned%20through%20child%20launch%2C%20or%20authenticate%20the%20bytes%20again%20when%20the%20child%20opens%20them.%0A%0A**How%20this%20was%20verified%3A**%20The%20authentication%20function%20returns%20%60governanceFile.filePath%60%2C%20and%20the%20resumed%20child%20later%20reads%20%60ULTRAFUZZ_DATA_GOVERNANCE_PATH%60%20by%20pathname%20without%20checking%20those%20bytes%20against%20the%20control%20seal.%0A%0A---%0A%0AFor%20each%20issue%20above%2C%20determine%20whether%20it%20is%20valid%20and%20should%20be%20fixed.%20If%20so%2C%20fix%20it%20directly.&repo=monad-developers%2Fultrafuzz&pr=1117&platform=github"><picture><source media="(prefers-color-scheme: dark)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixAllInClaudeDark.svg?v=6"><source media="(prefers-color-scheme: light)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixAllInClaude.svg?v=6">Fix All in Claude Code</picture></a>

<details><summary>Prompt To Fix All With AI</summary>

`````markdown
### Issue 1
packages/runtime/src/workflow-integrity.ts:490
**Governance Path Is Unpinned**

The continuation authenticates the governance file but returns only its pathname. Smithers opens that path later, after the authentication checks have ended. If a process can modify the run-owned snapshot tree, it can replace the file between authentication and use, causing the continuation to launch with policy bytes that do not match the sealed launch decision. Keep the authenticated file or snapshot directory pinned through child launch, or authenticate the bytes again when the child opens them.

**How this was verified:** The authentication function returns `governanceFile.filePath`, and the resumed child later reads `ULTRAFUZZ_DATA_GOVERNANCE_PATH` by pathname without checking those bytes against the control seal.

---

For each issue above, determine whether it is valid and should be fixed. If so, fix it directly.
`````

</details>

<sub>Reviews (1): Last reviewed commit: ["fix(runtime): retain authenticated pre-g..."](https://github.com/monad-developers/ultrafuzz/commit/a32149e2651d752c9ad8f02d84a8f09b92ded20f) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=61645326)</sub>

> Greptile also left **1 inline comment** on this PR.

**Context used:**

- Knowledge Base — [Execution runtime](https://app.greptile.com/monad-foudnation/-/custom-context/knowledge-base/monad-developers/ultrafuzz/-/docs/execution-runtime.md)
- Knowledge Base — [Runtime orchestration](https://app.greptile.com/monad-foudnation/-/custom-context/knowledge-base/monad-developers/ultrafuzz/-/docs/runtime-orchestration.md)

### CODE REVIEW

File: packages/runtime/src/workflow-integrity.ts [hunk 1]
```diff
@@ -439,6 +440,74 @@ export function workflowControlGeneration(projectRoot: string, layout: RunLayout
...
+  assertNoSymlinkComponents(layout.root, controlsRoot, "continuation governance snapshot");
+  if (
+    !sameWorkflowControlSnapshotChangeTokens(before, after) ||
+    workflowControlSnapshotStatToken(controls) !==
+      workflowControlSnapshotStatToken(fs.lstatSync(controlsRoot, { bigint: true })) ||
+    !readBoundedRegularFile(layout.root, sealPath, "workflow control seal").equals(sealContents)
+  ) {
+    throw new Error("sealed continuation governance changed while it was authenticated");
+  }
+  return governanceFile?.filePath;
```
[the remark this answers is not in the document]
Comment (aviggiano) [reply packages/runtime/src/workflow-integrity.ts:490]: Confirmed: the lexical policy path is not pinned through native child execution. This is a retained native execution boundary, and the PR description now says so explicitly; the change does not claim descriptor confinement.

Before this PR, native continuation deliberately selects the persisted mutable workflow and disables module confinement. See [main start-run.ts](https://github.com/monad-developers/ultrafuzz/blob/a0c6e1d4d247e40ba7b558334a6836acd4af1460/packages/runtime/src/start-run.ts#L519), [main native environment handling](https://github.com/monad-developers/ultrafuzz/blob/a0c6e1d4d247e40ba7b558334a6836acd4af1460/packages/runtime/src/smithers.ts#L6440), and the [existing continuation regression](https://github.com/monad-developers/ultrafuzz/blob/a0c6e1d4d247e40ba7b558334a6836acd4af1460/packages/runtime/test/runtime.test.ts#L14807), which verifies that replaced mutable workflow/agent bytes remain the sources selected for ordinary resume. The same-owner modification capability in this finding can already modify those checks under that contract.

A bounded local probe confirms policy replacement after authentication; it is a real limitation, not a false positive. Normal publication/refresh retains the read-only original generation, and no ordinary lifecycle operation replacing it was found. This stability repair authenticates and restores the missing original launch input while preserving the native continuation contract. Binding only policy through the existing full-snapshot descriptor machinery would require broader workflow/config and detached-child ownership changes, so those are outside this PR. Missing, changed or partial original authority at admission remains rejected and covered by tests.

**Comment by aviggiano** [comment]
CI on exact head `a32149e2651d752c9ad8f02d84a8f09b92ded20f`: integration shard 1 passed all 74 selected tests, with zero failures, and wrote its release-validation report. The subsequent upload sent 482 bytes successfully, then artifact finalization returned an intermediary HTTP 403. The job remains failed because its required report was not finalized.

The other lanes are still running. Once they finish, one retry of this specific failed shard is planned; their successful results will be retained. No source, workflow permissions, test budgets or required checks are being changed for this infrastructure failure. The full release gate must pass before merge.

**Comment by aviggiano** [comment]
The original run is complete. Shard 2 also passed every selected test (73 passed, zero failed), wrote an exit-zero/pass release report, and successfully finalized that artifact. The runner nevertheless cancelled the validation step at its 75-minute job deadline. Its 66 tests shared with the passing #1115 shard took 59.6 minutes versus 27.7 minutes there; seven added tests took 10.0 minutes and all passed. This shows broadly slower execution, without establishing its underlying cause.

One combined failed-job retry was accepted at 16:45:43 UTC on unchanged head `a32149e2651d752c9ad8f02d84a8f09b92ded20f`. Attempt 2 of run 34243952456 starts only the two diagnosed shards: shard 1's artifact-finalization 403 and shard 2's deadline cancellation after successful validation. The other successful lanes are carried forward, and the dependent aggregate will run again. No code or budget changes were made for this retry.

**Comment by aviggiano** [comment]
Final reviewed head `a32149e2651d752c9ad8f02d84a8f09b92ded20f` has passed all required gates. The single documented retry of the two diagnosed jobs passed: integration shard 1 in 32m10s, shard 2 in 30m42s, both including report upload; the final release gate passed at 17:18:32 UTC. Other successful lanes were retained. No product code, timing budget or workflow was changed for this retry.

Combined build, focused boundary regressions, full Bun contracts, and actual generated ordinary/refreshed reset runs were also verified. Current main plus this PR exactly matches the combined tested source tree. Merging the reviewed head normally; the documented historical-recovery and native-execution boundaries remain, and #1099 stays open.
